source/agora/test/PostTransactionResult.d

```diff
@@ -0,0 +1,61 @@
+/*******************************************************************************
+
+    Contains tests for the result of posting transactions which can be one
+    of these status: `Accepted``, `Duplicated`, and `Rejected`.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.PostTransactionResult;
+
+version (unittest):
+
+import agora.api.Validator;
+import agora.consensus.data.genesis.Test;
+import agora.crypto.Hash;
+import agora.crypto.Key;
+import agora.test.Base;
+import agora.utils.Test;
+
+import core.thread.osthread : Thread;
+
+/// test for posting transactions, waiting for the various results
+unittest
+{
+    TestConf conf = TestConf.init;
+    auto network = makeTestNetwork!TestAPIManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+    auto input_txs = GenesisBlock.txs.filter!(tx => tx.isPayment).array;
+    auto output_addr = WK.Keys.AA.address;
+
+    // create a transaction with the fee of 10000
+    auto tx = TxBuilder(input_txs[0], 0).feeRate(Amount(10_000))
+        .draw(Amount(100_000), [output_addr])
+        .sign();
+
+    // post the transaction, waiting for `Accepted`
+    auto result = nodes[0].postTransaction(tx);
+    assert(result == TransactionResult(TransactionStatus.Accepted));
+
+    // post a double spend transaction, waiting for `Duplicated`
+    result = nodes[0].postTransaction(tx);
+    assert(result == TransactionResult(TransactionStatus.Duplicated));
+
+    // post a transaction with no fee, waiting for `Rejected`
+    tx = TxBuilder(input_txs[0], 1).feeRate(Amount(0))
+        .draw(Amount(100_000), [output_addr])
+        .sign();
+    result = nodes[0].postTransaction(tx);
+    assert(result.status == TransactionResult.status.Rejected);
```

Comment:
We should also check the reason.

Reply:
I was thinking about checking the reason also but that could break the test when the content of the reason is changed even slightly. Let me think of a better way.

Reply:
It does not need to check the exact string but that it contains something relevant. If someone makes a breaking change to the message then the test will fail and it can be updated. It is better to have a test that can fail with a code change than no test at all.

Reply:
It's a simple way to test the reason. I applied it. Thanks.